File: rust/programs/worker_stake/src/instructions/worker/worker_stake.rs
use depin_core::{
    constants::BMB_MINT,
    utils::{
        account::{read_account_data, write_account_data},
        tokens::initialize_ata_if_needed,
    },
};
use solana_program::{
    account_info::{next_account_info, AccountInfo},
    entrypoint::ProgramResult,
    msg,
    program::invoke,
    program_error::ProgramError,
    pubkey::Pubkey,
};
use spl_token::instruction::transfer_checked;
use crate::{
    state::worker_stake_config::WorkerStakeConfig,
    types::WorkerStakeAccountType,
    utils::{validate_collection_authority, find_worker_stake_vault_pda},
};

pub fn process_worker_stake<'a>(
    program_id: &Pubkey,
    accounts: &'a [AccountInfo<'a>],
    amount: u64,
) -> ProgramResult {
    // Expected Accounts:
    // 0. [signer, writable] Worker collection update authority (payer for ATA)
    // 1. [readonly] Worker collection account (Metaplex Core collection)
    // 2. [writable] WorkerStakeConfig PDA
    // 3. [writable] Worker token account (source ATA for BMB)
    // 4. [readonly] Worker stake vault PDA (authority)
    // 5. [writable] Worker stake vault ATA (destination for BMB)
    // 6. [readonly] BMB mint
    // 7. [readonly] Token program
    // 8. [readonly] Associated token program
    // 9. [readonly] System program

    let account_info_iter = &mut accounts.iter();
    let collection_authority_account = next_account_info(account_info_iter)?;
    let worker_collection_account = next_account_info(account_info_iter)?;
    let worker_stake_config_account = next_account_info(account_info_iter)?;
    let worker_token_account = next_account_info(account_info_iter)?;
    let worker_stake_vault_pda = next_account_info(account_info_iter)?;
    let worker_stake_vault_ata = next_account_info(account_info_iter)?;
    let bmb_mint_account = next_account_info(account_info_iter)?;
    let token_program = next_account_info(account_info_iter)?;
    let associated_token_program = next_account_info(account_info_iter)?;
    let _system_program = next_account_info(account_info_iter)?;

    // Validate amount > 0
    if amount == 0 {
        msg!("Error: Amount must be greater than 0");
        return Err(ProgramError::InvalidArgument);
    }

    // Validate worker collection and authority
    validate_collection_authority(worker_collection_account, collection_authority_account)?;

    // Validate WorkerStakeConfig PDA
    let (config_pda, _bump) = WorkerStakeConfig::find_pda(program_id, worker_collection_account.key);
    if *worker_stake_config_account.key != config_pda {
        msg!("Error: WorkerStakeConfig account does not match expected PDA");
        return Err(ProgramError::InvalidArgument);
    }

    // Validate BMB mint
    if *bmb_mint_account.key != BMB_MINT {
        msg!("Error: Invalid BMB mint");
        return Err(ProgramError::InvalidArgument);
    }

    // Validate worker stake vault PDA
    let (expected_vault_pda, _vault_bump) = find_worker_stake_vault_pda(program_id, worker_collection_account.key);
    if *worker_stake_vault_pda.key != expected_vault_pda {
        msg!("Error: Worker stake vault PDA does not match expected address");
        return Err(ProgramError::InvalidArgument);
    }

    // Validate worker stake vault ATA matches derived address
    let expected_vault_ata = spl_associated_token_account::get_associated_token_address(
        worker_stake_vault_pda.key,
        bmb_mint_account.key,
    );
    if *worker_stake_vault_ata.key != expected_vault_ata {
        msg!("Error: Worker stake vault ATA does not match expected address");
        return Err(ProgramError::InvalidArgument);
    }

    // Initialize worker stake vault ATA if needed (lazy initialization)
    initialize_ata_if_needed(
        collection_authority_account,
        worker_stake_vault_pda,
        bmb_mint_account,
        worker_stake_vault_ata,
        token_program,
        associated_token_program,
    )?;

    // Transfer BMB from worker to worker stake vault
    let transfer_ix = transfer_checked(
        token_program.key,
        worker_token_account.key,
        bmb_mint_account.key,
        worker_stake_vault_ata.key,
        collection_authority_account.key,
        &[],
        amount,
        depin_core::constants::BMB_DECIMALS,
    )?;

    invoke(
        &transfer_ix,
        &[
            worker_token_account.clone(),
            bmb_mint_account.clone(),
            worker_stake_vault_ata.clone(),
            collection_authority_account.clone(),
            token_program.clone(),
        ],
    )?;

    // Load and update config
    let data = worker_stake_config_account.try_borrow_data()?;
    let mut config: WorkerStakeConfig = read_account_data(&data, WorkerStakeAccountType::WorkerStakeConfig)?;
    drop(data);

    config.total_staked = config.total_staked
        .checked_add(amount)
        .ok_or(ProgramError::ArithmeticOverflow)?;

    // Save updated config
    let mut data = worker_stake_config_account.try_borrow_mut_data()?;
    write_account_data(&mut data, WorkerStakeAccountType::WorkerStakeConfig, &config)?;

    msg!("Worker staked {} BMB. Total staked: {}", amount, config.total_staked);
    Ok(())
}

File: ts/packages/beamable-depin/src/features/worker_stake/worker-stake.ts
import {
    AccountRole,
    address,
    Address,
    Codec,
    Endian,
    getAddressEncoder,
    getStructCodec,
    getU64Codec,
    getProgramDerivedAddress
} from "gill";

import { BMB_MINT, SYSTEM_PROGRAM_ADDRESS, WORKER_STAKE_PROGRAM, WORKER_STAKE_VAULT_SEED } from "../../constants.js";
import { WorkerStakeInstruction } from "../../enums.js";
import { WorkerStakeConfigAccount } from "./worker-stake-config-account.js";
import { findAssociatedTokenPda, TOKEN_PROGRAM_ADDRESS, ASSOCIATED_TOKEN_PROGRAM_ADDRESS } from "@solana-program/token";

const addressEncoder = getAddressEncoder();

export interface WorkerStakeParams {
    amount: bigint;
}

export const WorkerStakeParamsCodec: Codec<WorkerStakeParams> = getStructCodec([
    ["amount", getU64Codec({ endian: Endian.Little })]
]);

export interface CreateWorkerStakeInput {
    collection_authority: Address;
    worker_collection: Address;
    amount: bigint;
}

export class WorkerStake {
    collection_authority: Address;
    worker_collection: Address;
    readonly params: WorkerStakeParams;

    constructor(input: CreateWorkerStakeInput) {
        this.params = {
            amount: input.amount
        };
        this.collection_authority = input.collection_authority;
        this.worker_collection = input.worker_collection;
    }

    private serialize(): Uint8Array {
        const inner = WorkerStakeParamsCodec.encode(this.params);
        return Uint8Array.of(WorkerStakeInstruction.WorkerStake, ...inner);
    }

    public async getInstruction() {
        const configPda = await WorkerStakeConfigAccount.findWorkerStakeConfigPDA(this.worker_collection);

        // Worker stake vault PDA
        const workerStakeVaultPda = await getProgramDerivedAddress({
            programAddress: WORKER_STAKE_PROGRAM,
            seeds: [WORKER_STAKE_VAULT_SEED, addressEncoder.encode(this.worker_collection)]
        });

        // ATA for worker stake vault
        const workerStakeVaultAta = await findAssociatedTokenPda({
            mint: BMB_MINT,
            owner: workerStakeVaultPda[0],
            tokenProgram: TOKEN_PROGRAM_ADDRESS
        });

        // ATA for collection authority's BMB tokens
        const workerTokenAccount = await findAssociatedTokenPda({
            mint: BMB_MINT,
            owner: this.collection_authority,
            tokenProgram: TOKEN_PROGRAM_ADDRESS
        });

        const accounts = [
            { address: this.collection_authority, role: AccountRole.WRITABLE_SIGNER },
            { address: this.worker_collection, role: AccountRole.READONLY },
            { address: configPda[0], role: AccountRole.WRITABLE },
            { address: workerTokenAccount[0], role: AccountRole.WRITABLE },
            { address: workerStakeVaultPda[0], role: AccountRole.READONLY },
            { address: workerStakeVaultAta[0], role: AccountRole.WRITABLE },
            { address: BMB_MINT, role: AccountRole.READONLY },
            { address: TOKEN_PROGRAM_ADDRESS, role: AccountRole.READONLY },
            { address: ASSOCIATED_TOKEN_PROGRAM_ADDRESS, role: AccountRole.READONLY },
            { address: SYSTEM_PROGRAM_ADDRESS, role: AccountRole.READONLY }
        ];

        return {
            programAddress: WORKER_STAKE_PROGRAM,
            accounts: accounts,
            data: this.serialize(),
        };
    }
}
